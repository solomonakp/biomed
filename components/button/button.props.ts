type ButtonBaseProps = {
    /**
     * True to show the loading state
     * @default false
     */
    loading?: boolean;
};

type HTMLButtonElementAttributes = React.ComponentPropsWithoutRef<'button'>;
export type ButtonProps = ButtonBaseProps & HTMLButtonElementAttributes;
