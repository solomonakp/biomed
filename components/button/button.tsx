import { Loader } from '@components/loader';
import { ButtonProps as Props } from './button.props';

export const Button = (props: Props) => {
    const { children, className, loading = false, disabled, ...rest } = props;

    return (
        <button
            className={`${className} px-5 py-1.5 font-bold  text-(--color-font-black) duration-150 bg-(--bg-button) rounded-lg hover:bg-indigo-100 active:bg-indigo-200 relative overflow-hidden inline-block transition focus:outline-black ease-in-out disabled:bg-gray-200 `}
            disabled={disabled || loading}
            role="button"
            {...rest}
        >
            {children}
            {loading && <Loader size="sm" variant="absolute" />}
        </button>
    );
};
