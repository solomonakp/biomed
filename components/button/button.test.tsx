import { render, screen, cleanup, fireEvent } from '@testing-library/react';
import { axe } from 'jest-axe';
import { Button } from './button';
import { ButtonProps as Props } from './button.props';

const defaultProps: Props = {
    onClick: jest.fn(),
    title: 'my button',
    role: 'button',
    type: 'button',
};

afterEach(cleanup);

const setUpButton = (props: Props = defaultProps) =>
    render(<Button {...props}>{defaultProps.title}</Button>);

describe('Button', () => {
    it('should the Button component be exported correctly', () => {
        expect(Button).toBeDefined();
    });
    test('should mount', () => {
        setUpButton();

        const button = screen.getByRole('button');
        expect(button).toHaveTextContent(defaultProps.title as string);

        expect(button).toBeInTheDocument();
    });

    test('should render the Button user would see', () => {
        const { asFragment } = setUpButton();

        expect(asFragment()).toMatchSnapshot();
    });

    test('should Button fire event when clicked', () => {
        const { onClick } = defaultProps;

        setUpButton();

        const button = screen.getByRole('button');

        fireEvent.click(button);

        expect(onClick).toHaveBeenCalled();
    });

    test('should Button show a loader while loading', async () => {
        const props = {
            ...defaultProps,
            loading: true,
        };
        setUpButton(props);

        const button = screen.getByRole('button');

        expect(button).toBeDisabled();
        expect(await screen.findByTestId(/loader/i)).toBeInTheDocument();
    });
    test('should Button not fail any accessibility test', async () => {
        const { container } = setUpButton();

        const result = await axe(container);

        expect(result).toHaveNoViolations();
    });
});
