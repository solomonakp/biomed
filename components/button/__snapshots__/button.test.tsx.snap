// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Button should render the Button user would see 1`] = `
<DocumentFragment>
  <button
    class="undefined px-5 py-1.5 font-bold  text-(--color-font-black) duration-150 bg-(--bg-button) rounded-lg hover:bg-indigo-100 active:bg-indigo-200 relative overflow-hidden inline-block transition focus:outline-black ease-in-out disabled:bg-gray-200 "
    role="button"
    title="my button"
    type="button"
  >
    my button
  </button>
</DocumentFragment>
`;
